src/transform/src/map_lifting.rs

```diff
@@ -398,7 +398,18 @@ impl LiteralLifting {
                 // lift literals from each input
                 let mut input_literals = Vec::new();
                 for input in inputs.iter_mut() {
-                    input_literals.push(self.action(input, gets));
+                    let (ok_literals, err_literals): (Vec<MirScalarExpr>, Vec<MirScalarExpr>) =
+                        self.action(input, gets)
+                            .into_iter()
+                            .partition(|f| f.is_literal_ok());
+
+                    // Do not propagate error literals beyond join inputs, since that may result
+                    // in them being propagated to other inputs of the join and evaluated when
+                    // they should not.
+                    if !err_literals.is_empty() {
+                        *input = input.take_dangerous().map(err_literals);
+                    }
+                    input_literals.push(ok_literals);
```

Comment:
Won't this potentially change the order of literals? Like, if I have `map([lit, err])` this would lift `lit` and push `err` back down, changing their order.

Reply:
oh, right, sorry :facepalm:

Reply:
I'm happy with the not-very-clever "if there is a single error do nothing", but otherwise I think we only want to strip from the end, or consider some permutation. But imo, this is probably a rare-ish case and optimizing the query may not be worth it for other than personal reasons. :D

Reply:
I guess the only option here is to do:

```
                // lift literals from each input
                let mut input_literals = Vec::new();
                for input in inputs.iter_mut() {
                    let literals = self.action(input, gets);

                    // Do not propagate error literals beyond join inputs, since that may result
                    // in them being propagated to other inputs of the join and evaluated when
                    // they should not.
                    if literals.iter().any(|l| l.is_literal_err()) {
                        *input = input.take_dangerous().map(literals);
                    } else {
                        input_literals.push(literals);
                    }
                }
```